[Test Proxy] Add test recording guide + migration guide (#35232)
Azure/azure-sdk-for-java: sdk/core/azure-core-test/README.md

- [Write or run tests](#write-or-run-tests)
    - [Set up test resources](#set-up-test-resources)
    - [Configure credentials](#configure-credentials)
    - [Start the test proxy server](#start-the-test-proxy-server)
    - [Write your tests](#write-your-tests)
    - [Configure live or playback testing mode](#configure-live-or-playback-testing-mode)
    - [Run and record tests](#run-and-record-tests)
    - [Run tests with out-of-repo recordings](#run-tests-with-out-of-repo-recordings)
    - [Sanitize secrets](#sanitize-secrets)
- [Troubleshooting](#troubleshooting)
- [Next steps](#next-steps)
- [Contributing](#contributing)

## Getting started

To use this package, add the following to your _pom.xml_.

[//]: # ({x-version-update-start;com.azure:azure-core-test;current})

```xml

<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-core-test</artifactId>
    <version>1.18.0</version>
</dependency>
```

[//]: # ({x-version-update-end})

## Key concepts

* Run tests in `Record` mode:
  To record means to intercept any HTTP request, store it in a file, then store the response received from the live
  resource that was originally targeted.
* Run tests in `Playback` mode:

>   To playback means to intercept any HTTP request and to respond it with the stored response of a previously recorded

Reviewer: ```suggestion
  To playback means to intercept any HTTP request and to respond  to it with the stored response of a previously recorded
```

  matching request.
* Run tests in `Live` mode:

>   To run live means to not intercept any HTTP request and to send them to the live resource.

Reviewer: ```suggestion
  To run live means to not intercept any HTTP request and to send them to the Azure service directly.
```

* Sanitize sensitive information:

>   Sensitive information means content like passwords, unique identifiers or personal information should be cleaned up

Reviewer: Do we have documentation on how to sanitize sensitive data? We should add a link to the documentation or elaborate here on how do it.

Author: Added in detailed section.

  from the recordings.
* [TestProxyTestBase][TestProxyTestBase.java]: Base test class that creates an `InterceptorManager` and enables the test
  to use [`test-proxy`][test-proxy-readme] for running the test. It either plays back test session data or records the
  test session.
* [InterceptorManager][InterceptorManager.java]: A class that keeps track of network calls by either reading the data
  from an existing test session record or recording the network calls in memory. Test session records are saved or read
  from "<i>.assets/{library-level}/src/test/resources/session-records/TestFileName.testName}.json</i>".
* [TestProxyRecordPolicy][TestProxyRecordPolicy.java]: Pipeline policy that records network calls using
  the [`test-proxy`][test-proxy-readme].
* [TestProxyPlaybackClient][TestProxyPlaybackClient.java]: HTTP client that plays back responses from the recorded data
  of session-record using test proxy.

## Write or run tests

Newer SDK tests utilize the [Azure SDK Tools Test Proxy][test-proxy-readme] to record and playback HTTP interactions.
To migrate from existing [TestBase][TestBase.java] to use the test proxy, or to learn more about using the test proxy,

> refer to the [test proxy migration guide][test-proxy-migration-guide].

Reviewer: This is already in the introduction section and we don't need to repeat this here.

    // method in TestProxyTestBase 
    @BeforeAll
    public static void setupTestProxy(TestInfo testInfo) {
        // Start the test proxy server
        testProxyManager.startProxy();
    }
}
```

For more details about how this starts up the test proxy, or the test proxy itself, refer to the
[test proxy migration guide][test-proxy-migration-guide].

## Write your tests

Each of the SDKs should include client sync and async testing in their `tests` directory (`sdk/{service}/{package}/tests`)
with the naming pattern `{ServiceName}ClientTest.java` and `{ServiceName}AsyncClientTest.java`.
The `{ServiceName}ClientTest` will be responsible for testing the synchronous client, and
the `{ServiceName}AsyncClientTest` will be responsible for testing the asynchronous client. 
The `{ServiceName}ClientTest` and the `{ServiceName}AsyncClientTest` both will extend the `{ServiceName}ClientTestBase` 
which then extends the `TestProxyTestBase` class.
The `{ServiceName}ClientTestBase` will be responsible for initializing the clients, preparing test data, registering
sanitizers/matchers etc. (in this example we use Tables SDK for the sake of demonstration):

> ```java

Author: For full code snippets, we will have to take dep on client libs which I was trying to avoid.

            playbackClient = interceptorManager.getPlaybackClient();
            tableServiceClientBuilder.httpClient(buildAssertingClient(playbackClient));
        } else {
            tableServiceClientBuilder.httpClient(buildAssertingClient(DEFAULT_HTTP_CLIENT));
            if (interceptorManager.isRecordMode()) {
                tableServiceClientBuilder.addPolicy(recordPolicy);
            }
        }
        if (!interceptorManager.isLiveMode()) {
            TestUtils.addTestProxyTestSanitizersAndMatchers(interceptorManager);
        }
        return tableServiceClientBuilder;
    }
}
```

## Configure live or playback testing mode

"Live" tests refer to tests that make requests to actual Azure resources. "Playback" tests require a recording for each
test; the test proxy will compare the requests/responses that would be made during each test with requests/responses in
the recording.

To run live tests, set the environment variable `AZURE_TEST_MODE` to `LIVE`.
To run tests in playback, set `AZURE_TEST_MODE` to `PLAYBACK` or leave it unset.

## Run and record tests

Set the environment variable `AZURE_TEST_MODE` to `RECORD` to run your test(s) in record mode.

> After tests finish running, there should folder called `src/test/resources/session-records` in your package

Reviewer: ```suggestion
After tests finish running, there should be a folder called `src/test/resources/session-records` in your package
```

> After tests finish running, there should folder called `src/test/resources/session-records` in your package

Reviewer: Is this still the case for a library that is already using the assets repo and we are just adding new tests to existing test classes?

Author: yes, once assets.json is present the new/updated files will be reflected in the respective `.assets` folder.

- **reset**: discards any pending changes to recordings, based on the tag pointer in `assets.json`.

After pushing your recordings, the `assets.json` file for your package will be updated to point to a new `Tag` that
contains the updates. Include this `assets.json` update in any pull request to update the recordings pointer in the
upstream repo.

### Sanitize secrets

The `.json` files created from running tests in record mode can include authorization details, account names, shared
access signatures, and other secrets. The recordings are included in our public GitHub repository, making it important
for us to remove any secrets from these recordings before committing them to the repository.

There are two primary ways to keep secrets from being written into recordings:

1. [Default sanitizers][default_sanitizers], similar to the use of the RecordingRedactor are already registered in the
   TestProxyUtils for default redactions.
2. Custom sanitizers can be added
   using `[TestProxySanitizer]`[test_proxy_sanitizer] & `interceptorManager.addSanitizers()` method for addressing
   specific service sanitization needs.

>    For example, registering a custom sanitizer for redacting the value of json key modelId from the response body looks

Reviewer: This uses Json Path and we should link to a document on how to use Json Path.
